experiments/compositions/composition/internal/controller/expander_reconciler.go

```diff
@@ -665,6 +706,147 @@ func (r *ExpanderReconciler) enqueueAllFromGVK(ctx context.Context, _ client.Obj
 	return reqs
 }
 
+func (r *ExpanderReconciler) reconcileDelete(ctx context.Context, logger logr.Logger, inputcr unstructured.Unstructured) (ctrl.Result, error) {
+	logger = logger.WithName("Delete")
+	if !controllerutil.ContainsFinalizer(&inputcr, finalizerName) {
+		return ctrl.Result{}, nil
+	}
+	planNN := types.NamespacedName{
+		Name:      r.InputGVR.Resource + "-" + inputcr.GetName(),
+		Namespace: inputcr.GetNamespace(),
+	}
+	plancr := compositionv1alpha1.Plan{}
+	if err := r.Client.Get(ctx, planNN, &plancr); err != nil {
+		logger.Error(err, "Unable to fetch Plan Object", "plan", planNN)
+		return ctrl.Result{}, err
+	}
+	annotations := plancr.GetAnnotations()
+	planLabels := plancr.GetLabels()
+	stageList, ok := annotations[stagesAnnotation]
+	if !ok {
+		err := fmt.Errorf("Plan is missing stage order annotation")
+		logger.Error(err, "Unable to fetch stage order from Plan", "Plan", planNN)
+		return ctrl.Result{}, err
+	}
+	stages := strings.Split(stageList, ",")
+	numFound := 0
+	for i := len(stages) - 1; i >= 0; i-- {
+		r.Recorder.Eventf(&inputcr, corev1.EventTypeNormal, "Delete", "Deleting objects for stage %s", stages[i])
+		nsList, ok := annotations[apply.ApplySetAdditionalNamespacesAnnotation]
+		if !ok {
+			err := fmt.Errorf("Plan is missing Namespace annotation")
+			logger.Error(err, "Unable to fetch Namespaces from Plan", "Plan", planNN)
+			return ctrl.Result{}, err
+		}
+		namespaces := strings.Split(nsList, ",")
+		namespaces = append(namespaces, inputcr.GetNamespace())
+		opts, err := deleteListOpts(stages[i], planLabels[apply.ApplySetParentIDLabel])
+		if err != nil {
+			logger.Error(err, "Error creating list options")
+			return ctrl.Result{}, err
+		}
+		gkList, ok := annotations[apply.ApplySetGKsAnnotation]
+		if !ok {
+			err := fmt.Errorf("Plan is missing GroupKind annotation")
+			logger.Error(err, "Unable to fetch GroupKinds from Plan", "Plan", planNN)
+			return ctrl.Result{}, err
+		}
+		gks := strings.Split(gkList, ",")
+		for _, gk := range gks {
+			parsedGK := schema.ParseGroupKind(gk)
+			mapping, err := r.RESTMapper.RESTMapping(parsedGK)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			n := 0
+			if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+				n, err = r.deleteNamespacedResources(ctx, logger, stages[i], mapping.Resource, namespaces, opts)
+			} else if mapping.Scope.Name() == meta.RESTScopeNameRoot {
+				n, err = r.deleteClusterResources(ctx, logger, stages[i], mapping.Resource, opts)
+			}
+			if err != nil {
+				logger.Error(err, "Error deleting resources", "GroupKind", gk)
+				r.Recorder.Eventf(&inputcr, corev1.EventTypeWarning, "Delete", "Failed deleting objects of GroupKind %q for stage %q: %v", gk, stages[i], err)
+				return ctrl.Result{}, err
+			}
+			numFound += n
+		}
+		if numFound > 0 {
```

Comment:
Logic here implies if a resource in a particular stage is not deleted (for whatever reason), the deletion of rest of the resources are blocked (fits the mental model). This is great if the delete pending resource eventually is removed (O(mins)). Iam thinking if we need to add a timeout (by looking at delete timestamp of the resource) to make progress if a resource is stuck in deletion. We can proceed to delete other resources after a timeout. We dont have to delete the plan/inputcr.

Reply:
we can do it in another PR ?